File: MATTHEW/leetcode/AIcoding/2.Gemini-2.0/python3_solutions/3331-find-subtree-sizes-after-changes.py
from collections import defaultdict

class Solution:
    def countSubtrees(self, n: int, edges: list[list[int]], queries: list[list[int]]) -> list[int]:
        adj = defaultdict(list)
        for u, v in edges:
            adj[u - 1].append(v - 1)
            adj[v - 1].append(u - 1)

        sizes = [0] * n

        def dfs(node, parent, blocked):
            sizes[node] = 1
            for neighbor in adj[node]:
                if neighbor != parent and neighbor != blocked:
                    sizes[node] += dfs(neighbor, node, blocked)
            return sizes[node]

        ans = []
        for u, v in queries:
            sizes = [0] * n
            dfs(u - 1, -1, v - 1)
            ans.append(sizes[u - 1])

        return ans

File: MATTHEW/leetcode/AIcoding/3.DeepSeek-v3/rust_solutions/3331-find-subtree-sizes-after-changes.rs
use std::collections::{HashMap, HashSet};

struct Solution;

impl Solution {
    pub fn subtree_sizes_after_changes(n: i32, edges: Vec<Vec<i32>>, queries: Vec<Vec<i32>>) -> Vec<i32> {
        let n = n as usize;
        let mut adj = vec![HashSet::new(); n];
        for edge in edges {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            adj[u].insert(v);
            adj[v].insert(u);
        }

        let mut parent = vec![0; n];
        let mut size = vec![0; n];
        let mut root = 0;
        Self::dfs(root, n, &adj, &mut parent, &mut size);

        let mut res = Vec::new();
        let mut current_adj = adj.clone();
        for query in queries {
            let u = query[0] as usize;
            let v = query[1] as usize;
            if current_adj[u].contains(&v) {
                current_adj[u].remove(&v);
                current_adj[v].remove(&u);
                if Self::is_parent(u, v, &parent) {
                    root = v;
                } else {
                    root = u;
                }
            } else {
                current_adj[u].insert(v);
                current_adj[v].insert(u);
                if Self::is_parent(u, v, &parent) {
                    root = u;
                } else {
                    root = v;
                }
            }
            Self::dfs(root, n, &current_adj, &mut parent, &mut size);
            res.push(size[root] as i32);
        }
        res
    }

    fn dfs(u: usize, parent_node: usize, adj: &[HashSet<usize>], parent: &mut [usize], size: &mut [usize]) {
        parent[u] = parent_node;
        size[u] = 1;
        for &v in adj[u].iter() {
            if v != parent_node {
                Self::dfs(v, u, adj, parent, size);
                size[u] += size[v];
            }
        }
    }

    fn is_parent(u: usize, v: usize, parent: &[usize]) -> bool {
        let mut node = v;
        while node != u && node != parent[node] {
            node = parent[node];
        }
        node == u
    }
}